packages/filesystem/src/browser/file-resource.ts

```diff
@@ -35,29 +37,55 @@ export class FileResource implements Resource {
     }
 
     async readContents(options?: { encoding?: string }): Promise<string> {
+        this.state = await this.doReadContents(options);
+        return this.state.content;
+    }
+    protected async doReadContents(options?: { encoding?: string }): Promise<FileResource.State> {
         try {
-            const { stat, content } = await this.fileSystem.resolveContent(this.uri.toString(), options);
-            this.stat = stat;
-            return content;
+            if (!await this.fileSystem.exists(this.uri.toString())) {
+                return FileResource.emptyState;
+            }
+            if (await this.isOutdated()) {
+                return this.fileSystem.resolveContent(this.uri.toString(), options);
+            }
+            return this.state;
         } catch {
-            return '';
+            return FileResource.emptyState;
+        }
+    }
+
+    protected async isOutdated(): Promise<boolean> {
+        if (!this.state.stat) {
+            return true;
         }
+        const stat = await this.fileSystem.getFileStat(this.uri.toString());
+        /*
+         * if timestamps equal we cannot assume that content has not been changed because of
+         * https://nodejs.org/en/docs/guides/working-with-different-filesystems/#timestamp-resolution
+         */
+        return stat.lastModification >= this.state.stat.lastModification;
     }
 
     async saveContents(content: string, options?: { encoding?: string }): Promise<void> {
+        const stat = await this.doSaveContents(content, options);
+        this.state = { stat, content };
+    }
+    protected async doSaveContents(content: string, options?: { encoding?: string }): Promise<FileStat> {
         const uri = this.uri.toString();
-        if (await this.fileSystem.exists(uri)) {
-            this.stat = this.stat || await this.fileSystem.getFileStat(this.uri.toString());
-            this.stat = await this.fileSystem.setContent(this.stat, content, options);
-        } else {
-            await this.fileSystem.createFile(uri, { content });
+        if (!await this.fileSystem.exists(uri)) {
```

Comment:
I would instead always try to `getFileStat` and catch the error when the file does not exist. If we have the error, then let's create the new file. So we could save one `exist` check each time.

This check is not atomic anyway, so even if it returns with `true` for the existence check, someone else could delete the file before we write it. What do you think?

Reply:
I know, but when we will always log an error in the backend, it is not nice

Reply:
And I have another tiny tiny feedback to this class in general; I would save a uri.toSrting() as a variable. Parsing the URI should not be expensive, but storing it is for sure cheaper.

Reply:
> I know, but then we will always log an error in the backend, it is not nice

I see your point. What about changing the API then? :)
`getFileStat(uri: string): Promise<FileStat>` => `getFileStat(uri: string): Promise<FileStat | undefined>`

Reply:
OK, let's put these aside until we fix the original issue. Sorry.

Reply:
let's file an issue for it, also for `setContents` maybe it can create a file if does not exist